pkg/dns/ibm/private/dnssvcs_provider.go

```diff
@@ -0,0 +1,273 @@
+		return fmt.Errorf("delete: invalid dns input data: %w", err)
+	}
+
+	listOpt := p.dnsService.NewListResourceRecordsOptions(p.config.InstanceID, zone.ID)
+	// Some dns records (e.g. wildcard record) have an ending "." character in the DNSName
+	DNSName := strings.TrimSuffix(record.Spec.DNSName, ".")
+
+	result, response, err := p.dnsService.ListResourceRecords(listOpt)
+	if err != nil {
+		if response != nil && response.StatusCode != http.StatusNotFound {
+			return fmt.Errorf("delete: failed to list the dns record: %w", err)
+		}
+	}
+	if result == nil {
+		return fmt.Errorf("delete: invalid result")
+	}
+	for _, target := range record.Spec.Targets {
+		for _, resourceRecord := range result.ResourceRecords {
+
+			var resourceRecordTarget string
+			rData, ok := resourceRecord.Rdata.(map[string]interface{})
+			if !ok {
+				return fmt.Errorf("delete: failed to get resource data: %v", resourceRecord.Rdata)
+			}
+			if *resourceRecord.Type == string(iov1.CNAMERecordType) {
+				resourceRecordTarget = fmt.Sprint(rData["cname"])
+			}
+			if *resourceRecord.Type == string(iov1.ARecordType) {
+				resourceRecordTarget = fmt.Sprint(rData["ip"])
+			}
+
+			if resourceRecordTarget == target && *resourceRecord.Name == DNSName {
+				delOpt := p.dnsService.NewDeleteResourceRecordOptions(p.config.InstanceID, zone.ID, *resourceRecord.ID)
+				delResponse, err := p.dnsService.DeleteResourceRecord(delOpt)
+				if err != nil {
+					if delResponse != nil && delResponse.StatusCode != http.StatusNotFound {
+						return fmt.Errorf("delete: failed to delete the dns record: %w", err)
+					}
+				}
+				if delResponse != nil && delResponse.StatusCode != http.StatusNotFound {
+					log.Info("deleted DNS record", "record", record, "zone", zone, "target", target)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateDNSServices validates that provider clients can communicate with
+// associated API endpoints by having each client list zones of the instance.
+func validateDNSServices(provider *Provider) error {
+	var errs []error
+
+	for _, zoneID := range provider.config.Zones {
+		getDnszoneOptions := provider.dnsService.NewGetDnszoneOptions(
+			provider.config.InstanceID,
+			zoneID)
+		_, _, err := provider.dnsService.GetDnszone(getDnszoneOptions)
+		if err != nil {
+			return fmt.Errorf("validateDNSServices: failed to get dns zone: %w", err)
+		}
+
+		listOpt := provider.dnsService.NewListResourceRecordsOptions(provider.config.InstanceID, zoneID)
+		listResult, response, err := provider.dnsService.ListResourceRecords(listOpt)
+		if err != nil {
+			if response != nil && response.StatusCode != http.StatusNotFound {
+				return fmt.Errorf("validateDNSServices: failed to list dns records: %w", err)
+			}
+		}
+		if listResult == nil {
+			return fmt.Errorf("validateDNSServices: invalid result")
+		}
+
+	}
+	return kerrors.NewAggregate(errs)
+}
+
+func (p *Provider) createOrUpdateDNSRecord(record *iov1.DNSRecord, zone configv1.DNSZone) error {
```

Comment:
*nit* does zone need to be pass-by-value here? Doesn't that do an unnecessary memory copy?
Same comment goes for `cis_provider.go` to be consistent.

Reply:
Changed it. (Side note: as I see this change is inconsistent with the way other platform providers do it.)

Reply:
@SzucsAti yea reviewing this again, I was asking the question, not necessarily suggesting it to be changed. Sorry I should have been more clear with my statement that I wasn't 100% sure.

I was originally thinking pointers are just overall better, but after reading https://medium.com/@meeusdylan/when-to-use-pointers-in-go-44c15fe04eac, it really only matters when there are large complex structures.

If you see this change as inconsistent, I'd recommend changing back. Sorry about that.

Reply:
Reverted the change.